File: beware-of-the-sheep/src/Entities/Animals/Animal.cpp
#include "pch.h"
#include "Animal.h"
#include "Entities/Plants/Plant.h"
#include "World.h"

namespace bots {

	Animal::Animal(World & world, const Point & position, unsigned int strength, unsigned int initiative) :
		Entity(world, position, strength, initiative)
	{}

	void Animal::action()
	{
		int vorh, norp;
		vorh = rand() % 2;
		norp = rand() % 2;
		
		if (vorh)
			move(0, norp ? -1 : 1);
		else
			move(norp ? -1 : 1, 0);
	}

	void Animal::collision(Entity & other)
	{
		std::string message;
		if (dynamic_cast<Plant*>(&other)) {
			other.kill();
			message = other.getClassName() + " was eaten by " + getClassName();
			m_world.addMessage(std::move(message));
			other.collision(*this);
		}
		else if (getKind() == other.getKind()) {
			moveToPrevPosition();
			Area::NeighboursArray arr{};
			unsigned int count = m_world.getFreeSpaceAround(getPosition(), arr);

			if (count > 0)
				m_world.spawnEntity(getKind(), arr[rand() % count]);
		}
		else if (other.dodgedAttack())
			moveToPrevPosition();
		if (other.getStrength() <= getStrength()) {
			other.kill();
			message = other.getClassName() + " was slain by " + getClassName();
			m_world.addMessage(std::move(message));
		}
		else {
			kill();
			message = getClassName() + " was slain by " + other.getClassName();
			m_world.addMessage(std::move(message));
		}
	}

} // namespace bots

File: src/entities/animal.py
from .entity import Entity
from .plant import Plant
import random

class Animal(Entity):
    def __init__(self, world, position, strength, initiative):
        super().__init__(world, position, strength, initiative)

    def action(self):
        v_or_h = random.randint(0, 1)
        n_or_p = random.randint(0, 1)

        if v_or_h == 1:
            self.move(0, -1 if n_or_p == 1 else 1)
        else:
            self.move(-1 if n_or_p == 1 else 1, 0)

    def collision(self, other):
        if isinstance(other, Plant):
            other.kill()
            self.world.addMessage(str(other) + ' was eaten by ' + str(self))
            other.collision(self)
        elif isinstance(other, self.__class__):
            self.moveToPrevPosition()
            free = self.world.getFreeSpaceAround(self.position)
            count = len(free)
            if count > 0:
                self.world.spawnEntity(self, free[random.randint(0, count - 1)])
        elif other.dodgedAttack(self.strength):
            self.moveToPrevPosition()
        elif other.strength <= self.strength:
            other.kill()
            self.world.addMessage(str(other) + ' was slain by ' + str(self))
        else:
            self.kill()
            self.world.addMessage(str(self) + ' was slain by ' + str(other))
